### PR TITLE
[SPARK-11171][SPARK-11237][SPARK-11241][ML] Try adding PMMLExportable to ML with KMeans

### PR DESCRIPTION
**Pull request by holdenk**
Add PMMLExportable to ML & use it for KMeans in ML pipeline. This is a bit different than the PMMLExportable in MLLib since the default trait doesn't go through a factory but instead depends on the model implementing it itself (and at the start that will mostly just be calling the parent model's method).
Still a WIP but wanted to see if anyone had comments about the slight change.

### CODE REVIEW

**Comment by holdenk** [comment]
cc @jkbradley

**Comment by holdenk** [comment]
re-ping @jkbradley

**Comment by holdenk** [comment]
re-ping @jkbradley

**Comment by holdenk** [comment]
Also maybe @dbtsai if you have a chance to look at this that would be great.

**Comment by holdenk** [comment]
So I've updated this against master @jkbradley

File: mllib/src/main/scala/org/apache/spark/ml/pmml/PMMLExportable.scala [hunk 1]
```diff
@@ -0,0 +1,90 @@
+/*
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/pmml/PMMLExportable.scala]: Seems this is a copy-paste of org.apache.spark.mllib.pmml, should we deprecate the mllib one, and use the new one in ml package?

Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/pmml/PMMLExportable.scala]: I think that might be good, the main difference is this avoids using the factory implementation that the MLLib API was.

Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/pmml/PMMLExportable.scala]: (Uses the same public facing API as per the JIRA discussion re: lack of complaints from users with old API)

**Comment by holdenk** [comment]
ping @jkbradley

**Comment by holdenk** [comment]
jenkins retest this please

**Comment by holdenk** [comment]
Jenkins retest this please

**Comment by holdenk** [comment]
jenkins retest this please

**Comment by MLnick** [comment]
@holdenk I took a very quick pass and the basics look ok. I will do a more detailed one soon.

I was thinking, just looking at this... what about seeing if we can do model export as a more pluggable approach? So, for example, in DataFrames we can do `df.write.format("json").save("...")`, and in ML we can do `kmeans.write.save("...")`.

So how about `kmeans.write.pmml.save("...")`? I'm not sure we need to do a full-blown generic implementation here (as for DataFrame) as it's a lot of overhead for not that much gain (essentially what is a "1st class citizen"? probably native Spark and PMML).

But something simple like making `PMMLExportable` interact with `MLWriter`:

```scala
  /** [[MLWriter]] instance for [[KMeansModel]] */
  private[KMeansModel] class KMeansModelWriter(instance: KMeansModel) extends MLWriter with PMMLWritable {

    private case class Data(clusterCenters: Array[Vector])

    override protected def saveImpl(path: String): Unit = {
      // Save metadata and Params
      DefaultParamsWriter.saveMetadata(instance, path, sc)
      // Save model data: cluster centers
      val data = Data(instance.clusterCenters)
      val dataPath = new Path(path, "data").toString
      sqlContext.createDataFrame(Seq(data)).repartition(1).write.parquet(dataPath)
    }

    override def pmml: MLWriter = new KMeansModelPMMWriter(instance)

  }
  
  private class KMeansModelPMMWriter(instance: KMeansModel) extends MLWriter {
    override protected def saveImpl(path: String): Unit = instance.parentModel.toPMML(sc, path)
  }
```

@jkbradley @mengxr @srowen thoughts?

**Comment by holdenk** [comment]
Any thoughts? I think since we are explicitly deprecating MLLib it would be nice to get something in for PMML export in the pipeline API.

**Comment by MLnick** [comment]
ping @jkbradley @mengxr @srowen - thoughts on my suggestion above?

I think the overall gist of this PR is fine, but it seems nicer to me to actually put a trait for PMMLExportable in the MLWriter for each supported model class, but I'm ok with either approach.

**Comment by srowen** [comment]
If the base change is just bringing the `PMMLExportable` trait to ML in a reasonable form, that seems fine. I might not know enough to know if I'm missing something, but that much seems fine.

Hm, wouldn't this require updates to more than `KMeansModel`? I'm probably missing something.

As to whether to continue to evolve the API -- sounds reasonable. On the other hand, I'm sort of concerned that there are 2-3 serialization mechanisms for models, not all of which are fully implemented, and implemented slightly differently across mllib and ml, and making further changes to how it works in ml seems like it's piling on some complexity. That said -- I myself would prefer more rather than less PMML support.

So the further changes -- I suppose I'm neutral on for this particular PR.

**Comment by vruusmann** [comment]
I've been experimenting with a standalone Spark ML Pipelines to PMML converter in recent days. The goal is to cover basic transformers (eg. `StringIndexer`, `OneHotEncoder`, `VectorAssembler`) and most popular tree-based models (eg. decision trees, random forest and GBMs).

The prototype should be available early next week. So far I've completely ignored the existing Spark's serialization infrastructure, including the `MLWritable` and `PMMLExportable` traits.

**Comment by mengxr** [comment]
I want to hold PMML support in spark.ml for Spark 2.1. We might need to discuss more about the potential coverage we can get, especially on pipeline support and feature transformers. It would be nice to proceed with that big picture in mind. @vruusmann Could you ping us back when you have the prototype ready?

On the interface side, it would be nice to have PMML (and perhaps JSON) support as part of MLWritable, e.g., `model.write.format("pmml").save("/path")`. We can discuss more about the APIs.

@vruusmann Are you familiar with PFA (http://dmg.org/pfa/index.html)? How does it compare to PMML?

**Comment by vruusmann** [comment]
The main difference between PMML and PFA is the abstraction level. PMML is a high-level language (more similar to modeling languages such as UML), where you're supposed to express the computation logic using a fixed vocabulary of ML domain-specific concepts. PFA is a medium- to low-level language (more similar to regular programming languages), where you can work around some PMML limitations by embedding arbitrary computation logic (eg. doing processing in loops).

I think that PMML's "high-levelness" and fixed vocabulary is a feature (not a bug that needs fixing by introducing another standard). The upside is that PMML processing and maintenance can be heavily automated. It is possible to load a PMML document, and do all sorts of transformations (eg. standardization, optimization) with it, and have a guarantee that the transformation output is functionally identical to the input. I'm specifically expanding my work in that direction - the Visitor API of the JPMML-Model library is the main tool here.

The difference in abstraction levels means that it's possible to translate PMML to PFA (but not the other way around). So, if you're thinking about adding PFA support to Apache Spark, then it might be sensible to first translate Spark ML Pipelines to PMML, and then translate PMML to PFA. The same PMML-to-PFA converter would be usable with all PMML documents regardless of the original modeling software (eg. R, Scikit-Learn, Apache Spark).

**Comment by holdenk** [comment]
@vruusmann that sounds interesting, I've been hoping to do some of that as well so do ping me once you've got something up.

**Comment by vruusmann** [comment]
Ping @srowen @mengxr @holdenk The first version of Spark ML pipelines to PMML converter is now available at: https://github.com/jpmml/jpmml-sparkml

However, be forewarned that I will be completely reworking the feature transforms-side of the codebase in the coming days. The goal is to have a generalized description of the "feature derivation graph", which could be shared between [JPMML-SkLearn](https://github.com/jpmml/jpmml-sklearn) and [JPMML-SparkML](https://github.com/jpmml/jpmml-sparkml) projects.

**Comment by vruusmann** [comment]
A thought about designing an interface for exporting ML solutions (exemplified using PMML, but should be generalizable to other data formats as well).

Namely, the export method should have a parameter, which gives the "best guess"-type of description of the associated feature schema:
```
interface PMMLExportable {
  void toPMML(OutputStream outputStream, FeatureSchema featureSchema);
}
```

It would be desirable to know the name, operational type (eg. continuous, categorical) and data type (eg. integer, float, double) of individual features, because this way it will be possible to generate more specialized/informative content.

Example one: description of regression terms:
```xml
<!-- Dummy feature schema -->
<NumericPredictor name="x33" coefficient="0.3401013140365875"/>
<NumericPredictor name="x34" coefficient="-1.2512420398555455"/>

<!-- Rich feature schema -->
<CategoricalPredictor name="Occupation" value="Executive" coefficient="0.3401013140365875"/>
<CategoricalPredictor name="Occupation" value="Farming" coefficient="-1.2512420398555455"/>
```

Example two: description of tree split conditions:
```xml
<!-- Dummy feature schema -->
<SimplePredicate field="x33" operator="equal" value="1.0"/>

<!-- Rich feature schema -->
<SimplePredicate field="Occupation" operator="equal" value="Executive"/>
```

Current `PMMLExportable` implementation classes are hard-coded to assume a dummy feature schema (ie. all features are continuous doubles that are named `x1`, `x2`, .., `x${numFeatures}`).

**Comment by holdenk** [comment]
@vruusmann Mind if I ask why the PMML export component your making is AGPL? If were looking to integrate this into Spark thats probably not a good license to use.

**Comment by vruusmann** [comment]
@holdenk The JPMML-SparkML library depends on AGPLv3-licensed libraries, which doesn't leave much choice.

I've just published the refactored feature mapping system. While implementing support for the `StandardScalerModel` transformation type, I came across a rather nasty issue related to the "tolerance" (actually, the lack thereof) of continuous tree split conditions. For example, on several occasions, the feature vector that was coming through `StandardScaler` was dispatched to the wrong tree branch, because there was a difference in the "16th decimal place of precision" between the computed feature value and the threshold value (eg. `x = 0.36367792586219525` does not satisfy the split condition `x <= 0.3636779258621952`).

**Comment by holdenk** [comment]
@vruusmann  That makes sense, although that means we can't include that code into Spark its self.

**Comment by holdenk** [comment]
So I guess I'm wondering what our plans for PMML look like - I'm happy to update this or go in the direction @MLnick suggested if thats what we want?

**Comment by holdenk** [comment]
@mengxr One downside of going the generic writer way is we loose compile time checking as to if the requested format is supported.

**Comment by MLnick** [comment]
@holden that's true for the fully generic approach. But `DataFrameWriter` for example exposes `json` as a shortcut (type-safe in a way) for `format("json")`.

I think we can achieve something similar here, but since each impl of `MLWriter` is different for each model, I think we can have a `PMML` trait attached to those that support it, enabling the "type-safe" approach: `model.write.pmml.save("/path")`.

`model.write.save("/path")` does the default built-in format, `model.write.pmml.save("/path")` does pmml for those models that actually support it using the trait. For generics, it's possible to do `model.write.format("pmml").save(...)` and it would then fail at runtime if not supported, while `model.write.format("my.custom.format").save(...)` could allow plugging in writers similar to the datasource API...

just thoughts, obviously more work will be required to see if it is feasible in practice.

**Comment by holdenk** [comment]
@MLnick I like that idea! I've done the shortcut approach on the root (since thats the shortcut we already had) - but I'll look at adding the shortcut on the MLWriter its self as well. Thanks!

**Comment by holdenk** [comment]
HiveSubmit failures seem unrelated, jenkins retest this please.

**Comment by holdenk** [comment]
Now that 2.0 RC5 has passed - maybe it would be an OK time to revisit this? I think it can also be an ok base for adding more formats if we need to.

**Comment by holdenk** [comment]
ping @MLnick now that 2.0 is out for awhile.

**Comment by holdenk** [comment]
Ping @jkbradley / @mlnick - we are coming up on a year since we told people we would add PMML export to Spark in about a year so I figured I'd try and update this again. I'll update this PR  tonight against the latest master.

**Comment by holdenk** [comment]
@MLNick: Do you have the bandwith to revisit this? I'm open to refactoring to a more plug-gable approach if we've got the review bandwidth for it.

**Comment by holdenk** [comment]
Closing this in favor of the approach taken in https://github.com/apache/spark/pull/19876

**Pull request closed without merging**